### DISCUSSION TITLE
The correct approach to add hover-out transitions

### DISCUSSION THREAD

**Discussion by vilebbit** [tailwindlabs/tailwindcss#9769] [Help] [answered]
Hello, I'm a newly to tailwind and working on a project based on it but facing an intractable situation, any advice would be appreciated.

I know one can add hover transitions to an element by doing something like this:

```html
<div className='bg-zinc-300 transition-colors duration-300 hover:bg-amber-300'>
  my contents
</div>
```

In fact, this works very well. 
But things go a little differently while it has a "dark" variant: 

```html
<div className='bg-zinc-300 transition-colors duration-300 hover:bg-amber-300 dark:bg-zinc-700 dark:hover:bg-amber-700'>
  my contents
</div>
```

The above codes still work good while moving cursor on the element. However, while changing theme between light and dark, I got a brief transition time of the color changing as well which makes it looks somehow abrupt. 

What I truly want is a transition only takes effect while hovering events happened. As for when changing themes, I need it to transform immediately without any transitions. 

By searching [the document](https://tailwindcss.com/docs/transition-duration#hover-focus-and-other-states), I tried the approach from it:

```html
<div className='bg-zinc-300 transition-colors duration-0 hover:duration-300 hover:bg-amber-300 dark:bg-zinc-700 dark:hover:bg-amber-700'>
  my contents
</div>
```

However, by doing like this, transitions only take effect while hovering-on to the element, leaving hover-out transition blank. 
I also have tried some other ways like adding `[&:not(:hover)]:duration-300` but none of them make sense. 

So, is there a way to add hover-out transitions? Or did I get the wrong approach to prevent transitions occur during theme changing?

**Comment by wongjn** [comment, marked as answer]
If you are using the class `darkMode` strategy, you could add the `duration-0` and remove the `duration-300` class a frame before you add/remove the dark class, and then revert the `duration-` class changes a frame afterwards. You'd do these frame timings via [`requestAnimationFrame()`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame):
```js
const darkModeChange = () => {
  element.classList.remove('duration-300');
  element.classList.add('duration-0');
  // On next frame:
  requestAnimationFrame(() => {
    toggleDarkMode();
    // On next frame:
    requestAnimationFrame(() => {
      element.classList.add('duration-300');
      element.classList.remove('duration-0');
    });
  });
};
```
Or if you are using the media query `darkMode` strategy, we would not be able to know a frame early when the dark mode kicks in, so instead, we would have to keep `duration-0` on the element by default and only add `duration-300` once the user hovers on the element to transition when their mouse leaves:
```html
<div class='bg-zinc-300 transition-colors duration-0 hover:duration-300 hover:bg-amber-300 dark:bg-zinc-700 dark:hover:bg-amber-700'>
  my contents
</div>
<script>
  const element = document.querySelector('div');

  const remove = ({ target }) => {
    // `transitionend` bubbles, so make sure it is the element
    // we attached the listener to.
    if (target === element) {
      element.addClass('duration-0');
      element.remove('duration-300');
      element.removeEventListener('transitionend', remove);
    }
  };
 
  element.addEventListener('mouseenter', () => {
    // Explicitly remove the `transitionend` listener if we hover
    // the element again, in case we hovered while the `mouseleave`
    // transition was still playing.
    element.removeEventListener('transitionend', remove);
  
    element.removeClass('duration-0');
    element.addClass('duration-300');
  });
  element.addEventListener('mouseleave', () => {
    // Attach callback to remove the `duration-300` class once the
    // transition has ended.
    element.addEventListener('transitionend', remove);
  });
</script>
```

**Reply by vilebbit** [reply]
> If you are using the class `darkMode` strategy, you could add the `duration-0` and remove the `duration-300` class a frame before you add/remove the dark class, and then revert the `duration-` class changes a frame afterwards. You'd do these frame timings via ...


Thanks, this does solve the problem but, I wonder is it really necessary to do something this complicated to achieve the feature. 
Additionally, I have dozen of components come with transitions, it's not realistic to write down all of them.

**Reply by wongjn** [reply]
It is necessary to do something like this to achieve the feature. To make it work more "globally", you could use an arbitrary variant to target every element in the DOM:
```js
const darkModeChange = () => {
  document.documentElement.classList.add('[&_*]:!duration-0');
  // On next frame:
  requestAnimationFrame(() => {
    toggleDarkMode();
    // On next frame:
    requestAnimationFrame(() => {
      document.documentElement.classList.remove('[&_*]:!duration-0');
    });
  });
};
```

**Reply by adamwathan** [reply]
+1, this is basically what we do on our [Syntax template](https://syntax.tailwindui.com/) for example:

```js
  function updateThemeWithoutTransitions(savedTheme) {
    updateTheme(savedTheme)
    document.documentElement.classList.add('[&_*]:!transition-none')
    window.setTimeout(() => {
      document.documentElement.classList.remove('[&_*]:!transition-none')
    }, 0)
  }
```

**Reply by vilebbit** [reply]
Okay... If that's the case, I think now I understand why most of websites which have theme changing features don't come with transitions. 
Many thanks to all of you for your advices.
